examples/gno.land/r/gov/dao/v3/impl/types.gno

```diff
@@ -0,0 +1,198 @@
+package impl
+
+import (
+	"strings"
+
+	"gno.land/p/demo/avl"
+	"gno.land/p/demo/ufmt"
+	"gno.land/r/gov/dao"
+	"gno.land/r/gov/dao/v3/memberstore"
+)
+
+type Law struct {
+	Supermajority float64
+}
+
+func (l *Law) String() string {
+	return ufmt.Sprintf("This law contains the following data:\n\n- Supermajority: %v%%", l.Supermajority)
+}
+
+// ProposalsStatuses contains the status of all the proposals indexed by the proposal ID.
+type ProposalsStatuses struct {
+	*avl.Tree // map[int]*proposalStatus
+}
+
+func NewProposalsStatuses() ProposalsStatuses {
+	return ProposalsStatuses{avl.NewTree()}
+}
+
+func (pss ProposalsStatuses) GetStatus(id dao.ProposalID) *proposalStatus {
+	pids := ufmt.Sprintf("%v", int(id))
+	psv, ok := pss.Get(pids)
+	if !ok {
+		return nil
+	}
+
+	ps, ok := psv.(*proposalStatus)
+	if !ok {
+		panic("ProposalsStatuses must contains only proposalStatus types")
+	}
+
+	return ps
+}
+
+type proposalStatus struct {
+	YesVotes memberstore.MembersByTier
+	NoVotes  memberstore.MembersByTier
+	AllVotes memberstore.MembersByTier
+
+	Accepted bool
+	Denied   bool
+
+	DeniedReason string
+
+	TiersAllowedToVote []string
+
+	TotalPower float64 // TotalPower is the power of all the members existing when this proposal was created.
+}
+
+func newProposalStatus(allowedToVote []string) *proposalStatus {
+	return &proposalStatus{
+		YesVotes: memberstore.NewMembersByTier(),
+		NoVotes:  memberstore.NewMembersByTier(),
+		AllVotes: memberstore.NewMembersByTier(),
+
+		TiersAllowedToVote: allowedToVote,
+
+		TotalPower: memberstore.Get().GetTotalPower(),
+	}
+}
+
+func (ps *proposalStatus) YesPercent() float64 {
+	var yp float64
+
+	memberstore.Tiers.Iterate("", "", func(tn string, ti interface{}) bool {
```

Comment:
can you compute and cache this when people vote so we can avoid this readonly iteration?